### PR TITLE
Multi-Label Joint Intent Slot Classification

### PR DESCRIPTION
**Pull request by chenrichard10**
# What does this PR do ?

Multi-Label Intent Model for Joint Intent Slot Classification Task. 

**Collection**: [Note which collection this PR will affect]

This PR will involves changes in collections/nlp within NeMo.

# Changelog 
- New loss function added to collections/common/losses (BCEWithLogitsLoss)
- Multi-Label Classification Report added to collections/nlp/metrics/classification_report
- Multi-Label Classification Model, Descriptor, Dataset files added in similar folders to intent_slot_classification
- Data Conversion and Augmentation scripts added to example/nlp/multi_label_intent_slot_classification Multi-Label Training Tutorial added to https://github.com/NVIDIA/NeMo/blob/main/tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb
- Added Weight Calculation Function for BCEWithLogitsLoss (lines 512-537) and Multi Label Stats function (lines 201-242) in nlp/data/data_utils/data_preprocessing.py 

# Usage

Refer to https://github.com/NVIDIA/NeMo/blob/main/tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb for training. To run the tutorial you will need to change github string link name from NVIDIA to chenrichard10 if the PR has not been approved. 

A simple script that can be used to train the model locally can be with  
 
```python
from nemo.collections import nlp as nemo_nlp
from nemo.utils.exp_manager import exp_manager
from nemo.utils import logging
from nemo.collections.nlp.parts.utils_funcs import tensor2list

import os
import time

import torch
import pytorch_lightning as pl

from omegaconf import OmegaConf

HOME_DIR=""

# directory with data converted to nemo format
data_dir = os.path.join(HOME_DIR, "new_multiatis")

run_name = "test"
config_file = ""
config = OmegaConf.load(config_file)
config.model.data_dir = data_dir
config.model.validation_ds.prefix = "dev"
config.model.test_ds.prefix = "test"
config.model.class_balancing = "weighted_loss"
config.model.head.num_output_layers = 2
config.trainer.max_epochs = 40


# checks if we have GPU available and uses it
cuda = 1 if torch.cuda.is_available() else 0
config.trainer.gpus = cuda

# config.trainer.precision = 16 if torch.cuda.is_available() else 32
config.trainer.precision = 32

# for mixed precision training, uncomment the line below (precision should be set to 16 and amp_level to O1):
# config.trainer.amp_level = 'O1'

# remove distributed training flags
config.trainer.accelerator = None

# early_stop_callback = EarlyStopping(monitor='intent_f1', min_delta=1e-1, patience=10, verbose=True, mode='max')

trainer = pl.Trainer(**config.trainer)
config.exp_manager.exp_dir = os.path.join(HOME_DIR, "output/" + run_name)
config.exp_manager.create_checkpoint_callback = True
config.exp_manager.version = time.strftime('%Y-%m-%d_%H-%M-%S')

exp_dir = exp_manager(trainer, config.get("exp_manager", None))
model = nemo_nlp.models.MultiLabelIntentSlotClassificationModel(config.model, trainer=trainer)
trainer.fit(model)
```

**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
Available model has yet to be trained and uploaded on NGC. Section will be updated once a model has been trained.

### CODE REVIEW

File: nemo/collections/nlp/models/multi_label_intent_slot_classification/multi_label_intent_slot_classification_model.py [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+            parts = [1 if label in parts else 0 for label in range(len(self.cfg.data_desc.intent_labels))]
+            metrics_labels.append(parts)
+
+        # Retrieve class probabilities for each sentence
+        intent_probabilities = self.prediction_probabilities(sentences, test_ds)
+
+        metrics_dict = {}
+        # Find optimal logits rounding threshold for intents
+        for i in np.arange(0.5, 0.96, 0.01):
+            predictions = tensor2list((intent_probabilities >= i).int())
```
Comment (carolmanderson) [inline nemo/collections/nlp/models/multi_label_intent_slot_classification/multi_label_intent_slot_classification_model.py:289]: This line will also need to be modified if you convert the torch tensor to a numpy array in `prediction_probabilities`. You can use numpy `.tolist()`

File: nemo/collections/nlp/models/multi_label_intent_slot_classification/multi_label_intent_slot_classification_model.py [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+
+        logging.info(
+            f'Maximum Threshold for F1-Score: {max_f1_score}, [Precision, Recall, F1-Score]: {metrics_dict[max_f1_score]}'
+        )
+        logging.info(
+            f'Maximum Threshold for Precision: {max_precision}, [Precision, Recall, F1-Score]: {metrics_dict[max_precision]}'
+        )
+        logging.info(
+            f'Maximum Threshold for Recall: {max_recall}, [Precision, Recall, F1-Score]: {metrics_dict[max_recall]}'
+        )
```
Comment (carolmanderson) [inline nemo/collections/nlp/models/multi_label_intent_slot_classification/multi_label_intent_slot_classification_model.py:307]: I found this logging output a little confusing. I think it would be clearer if it said "Best threshold" instead of "Maximum threshold".

File: nemo/collections/nlp/models/multi_label_intent_slot_classification/multi_label_intent_slot_classification_model.py [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+
+        if metrics_dict[max_f1_score][2] > self.max_f1:
+            self.max_f1 = metrics_dict[max_f1_score][2]
+            self.threshold = max_f1_score
```
Comment (carolmanderson) [inline nemo/collections/nlp/models/multi_label_intent_slot_classification/multi_label_intent_slot_classification_model.py:311]: Since you set a threshold here, I think you should also add a line of logging to tell the user that. Something like "Setting threshold to <whatever>"

File: tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb [hunk 4]
```diff
@@ -556,17 +556,251 @@
...
+    "Before using our model, we find the probability rounding threshold that gives us the best results on the validation set. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "eval_model.optimize_threshold(config.model.test_ds)"
```
Comment (carolmanderson) [inline tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb:767]: After running `optimize_threshold`, I think it would be good to add a cell where you call `eval_model.threshold` so the user can see that the "best" threshold, based on F1 scores, was set as a model attribute. Also please explain that in the notebook.

File: examples/nlp/multi_label_intent_slot_classification/conf/multi_label_intent_slot_classification_config.yaml [hunk 5]
```diff
@@ -0,0 +1,111 @@
+# Intent and Slot classification with pretrained BERT models
+
+trainer:
+  gpus: 2 # the number of gpus, 0 for CPU
```
Comment (carolmanderson) [inline examples/nlp/multi_label_intent_slot_classification/conf/multi_label_intent_slot_classification_config.yaml:4]: Set the default number of GPUs to 1 since most people won't have more than one. I don't think Colab is offering more than one either.

**Review by carolmanderson — CHANGES_REQUESTED** [review]
Thanks for this valuable contribution to NeMo! I haven't yet reviewed all scripts, but there's plenty to work on based on my comments so far. 

To my mind it would make more sense to have the `multi_label_intent_slot_classification_model.py` sit inside the `models/intent_slot_classification` folder, instead of in its own folder. And similar for the dataset and data descriptor. I know we discussed this before, and maybe there was some good reason you decided not to do that.

File: tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb [hunk 6]
```diff
@@ -555,6 +557,253 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Optimizing Threshold\n",
+    "\n",
+    "As mentioned above, when classifiying a given query such as `show all flights and fares from denver to san francisco`, our model checks whether each individual intent would be suitable. Before assigning the final labels for a query, the model assigns a probability an intent matches the query. For example, if our dict.intents had 5 different intents, then the model could output for a given query \\[0.52, 0.38, 0.21, 0.67. 0.80\\] where each value represents the probability that query matches that particular intent. \n",
+    "\n",
```
Comment (carolmanderson) [inline tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb:755]: Great explanation added here with the example probabilities.

File: nemo/collections/nlp/data/data_utils/data_preprocessing.py [hunk 7]
```diff
@@ -461,9 +504,36 @@ def get_freq_weights(label_freq):
...
+
+    for label_id, class_values in label_freq.items():
+        positive_class = class_values[1]
+        negative_class = class_values[0]
+
+        if positive_class == 0:
+            weights[label_id] = 0
+
+        else:
+            weights[label_id] = float(negative_class) / float(positive_class)
```
Comment (carolmanderson) [inline nemo/collections/nlp/data/data_utils/data_preprocessing.py:532]: I think this should be negative/total   (i.e. negative/(negative + positive)

File: nemo/collections/nlp/data/data_utils/data_preprocessing.py [hunk 8]
```diff
@@ -461,9 +504,36 @@ def get_freq_weights(label_freq):
     for lf in label_freq.values():
         total_size += lf
     weighted_slots = {label: (total_size / (len(label_freq) * freq)) for label, freq in label_freq.items()}
+
     return weighted_slots
 
 
+def get_freq_weights_bce_with_logits_loss(label_freq):
+    """
+    Calculate positive class weights in regards to BCEWithLogitsLoss
```
Comment (carolmanderson) [inline nemo/collections/nlp/data/data_utils/data_preprocessing.py:513]:
```suggestion
    Calculate positive class weights to be passed to BCEWithLogitsLoss
```

File: nemo/collections/common/losses/bce_logits_loss.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        return {"loss": NeuralType(elements_type=LossType())}
+
+    def __init__(
+        self, logits_ndim=2, weight=None, reduction="mean", ignore_index=-100, pos_weight=None,
+    ):
+        """
+        Args:
+            logits_ndim (int): number of dimensions (or rank) of the logits tensor
+            weight (list): list of rescaling weight given to each class
+            reduction (str): type of the reduction over the batch
```
Comment (carolmanderson) [inline nemo/collections/common/losses/bce_logits_loss.py:54]: `weight` and `ignore_index` are currently not used. Does it make sense to have both `weight` and `pos_weight`?

File: nemo/collections/common/losses/bce_logits_loss.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        """
+        if pos_weight is not None and not torch.is_tensor(pos_weight):
+            pos_weight = torch.FloatTensor(pos_weight)
+        # Think about ignore index
```
Comment (carolmanderson) [inline nemo/collections/common/losses/bce_logits_loss.py:58]: Funny inline comment....have you thought about it?

It looks like that argument isn't being used. Should it be in the `super` call?

File: examples/nlp/multi_label_intent_slot_classification/data/convert_datasets.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+    multi_intent_df = pd.DataFrame(intent_labels)
+    multi_intent_df.to_csv(f'{target_dir}/{mode}.tsv', sep='\t', index=False)
+
+
+def convert_intent_dictionary(source_dir: str, target_dir: str) -> None:
+    """
+    Converts original intent dictionary containing labels that represented multiple labels into 
+    dictionary with only single labels. Example: if index 5 was referring to label "a+b", it is no longer 
+    a label in the new intent dictionary. Only labels "a" and "b" are included within the new dictionary
```
Comment (carolmanderson) [inline examples/nlp/multi_label_intent_slot_classification/data/convert_datasets.py:75]: Nice explanation

File: nemo/collections/common/losses/bce_logits_loss.py (hunk 10 above)
Comment (chenrichard10) [reply nemo/collections/common/losses/bce_logits_loss.py:58]: BCEWithLogitsLoss does not have a ignore index argument,  so I will remove that

File: nemo/collections/common/losses/bce_logits_loss.py (hunk 9 above)
Comment (chenrichard10) [reply nemo/collections/common/losses/bce_logits_loss.py:54]: The original BCEWithLogitsLoss seems to have weight and pos_weight parameters (https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html) I think it would be good to have both and pass them in super().

File: nemo/collections/nlp/data/data_utils/data_preprocessing.py (hunk 7 above)
Comment (chenrichard10) [reply nemo/collections/nlp/data/data_utils/data_preprocessing.py:532]: I think from https://pytorch.org/docs/1.9.1/generated/torch.nn.BCEWithLogitsLoss.html it says that the pos_weight for a class should be negative / positive example (The paragraph right above "Examples:"

File: nemo/collections/nlp/models/intent_slot_classification/multi_label_intent_slot_classification_model.py [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+        )
+
+        if metrics_dict[max_f1_score][2] > self.max_f1:
+            self.max_f1 = metrics_dict[max_f1_score][2]
+
+            logging.info(f'Setting Threshold to: {max_f1_score}')
+
+            self.threshold = max_f1_score
+
+    def predict_from_examples(
```
[the remark this answers is not in the document]
Comment (chenrichard10) [reply nemo/collections/nlp/models/intent_slot_classification/multi_label_intent_slot_classification_model.py:357]: I don't think you can set self.threshold as a default argument, so I set it as 0.5 at the moment.

File: tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb [hunk 13]
```diff
@@ -555,6 +557,253 @@
...
+   "metadata": {},
+   "source": [
+    "## Evaluation"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To see how the model performs, we can evaluate the performance of the trained model on a test data file. Here we would load the best checkpoint (the one with the lowest validation loss) and create a model (eval_model) from the checkpoint. We will use the same trainer for testing. We will need to use the file specified in the output during the training process."
```
[the remark this answers is not in the document]
Comment (carolmanderson) [reply tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb:729]: I hadn't noticed this before, but there are actually two locations in the notebook where this needed to be changed -- above for the single-label model, and here for the multi-label model.

File: nemo/collections/nlp/models/intent_slot_classification/multi_label_intent_slot_classification_model.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (carolmanderson) [reply nemo/collections/nlp/models/intent_slot_classification/multi_label_intent_slot_classification_model.py:359]: Ah yeah, you're right of course. The usual workaround for this situation is to set the default as None and then replace it with the object attribute inside the method. I think this will be better than using 0.5, which might be unexpected if a user has just gone through the threshold finding and thinks the threshold is now going to be used.

File: examples/nlp/intent_slot_classification/conf/multi_label_intent_slot_classification_config.yaml [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    prefix: dev
+    batch_size: 32
+    shuffle: false
+    num_samples: -1
+    num_workers: 8
+    drop_last: false
+    pin_memory: false
+
+  test_ds:
+    prefix: test
```
Comment (carolmanderson) [inline examples/nlp/intent_slot_classification/conf/multi_label_intent_slot_classification_config.yaml:49]: Should this be changed to "dev" so that it will work easily with the multi-atis dataset? Alternatively you can pass this in as a command line arg in your docstring example.

File: nemo/collections/nlp/data/data_utils/data_preprocessing.py (hunk 7 above)
Comment (carolmanderson) [reply nemo/collections/nlp/data/data_utils/data_preprocessing.py:537]: Ah yes, you're right, of course.

**Review by carolmanderson — CHANGES_REQUESTED** [review]
Almost done! I reopened a couple of "resolved" conversations because the changes didn't get committed. Added a few more comments/suggestions too, but this is looking good.

File: nemo/collections/nlp/data/intent_slot_classification/multi_label_intent_slot_classification_dataset.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    and slot classification with pretrained model.
+
+    Converts from raw data to an instance that can be used by
+    NMDataLayer.
+
+    Args:
+        input_file: file containing sentences + labels. The first line is header (sentence [tab] label)
+            each line should be [sentence][tab][label] where label can be multiple labels separated by a comma
+        slot_file: file containing slot labels, each line corresponding to slot labels for a sentence in input_file. No header.
+        max_seq_length: max sequence length minus 2 for [CLS] and [SEP]
```
Comment (carolmanderson) [inline nemo/collections/nlp/data/intent_slot_classification/multi_label_intent_slot_classification_dataset.py:39]: num_intents should be included in the args

File: nemo/collections/nlp/metrics/classification_report.py [hunk 16]
```diff
@@ -105,10 +115,11 @@ def update(self, predictions: torch.Tensor, labels: torch.Tensor):
         self.fp += fp
         self.num_examples_per_class += num_examples_per_class
 
-    def compute(self):
+    def compute(self) -> None:
```
Comment (carolmanderson) [inline nemo/collections/nlp/metrics/classification_report.py:118]: I don't think this should be None since there are return values.

**Review by carolmanderson — CHANGES_REQUESTED** [review]
Looks great! I've just requested two tiny changes.
